File: hackerrank/contests/week-of-code21/n-letter.py
from math import acos, sqrt, pi
from itertools import permutations

def orientation(a, b, c):
	"""
	-1 => counter clock wise
	1 => clockwise
	"""
	val = (b[1]-a[1])*(c[0]-b[0]) - (b[0]-a[0])*(c[1]-b[1])
	if val == 0: return 0
	if val > 0:
		return 1
	return -1

def isangle(a, b, c):
	angle = acos(float((a[0]-b[0])*(c[0]-b[0]) + (a[1]-b[1])*(c[1]-b[1]))/
					float(sqrt((b[0]-a[0])** 2 + (b[1]-a[1])**2) * sqrt((b[0]-c[0])** 2 + (b[1]-c[1])**2)))
	angle *= 180
	angle /= pi
	if angle > 90: return False
	return True


points = []

n = input()

for a in xrange(n):
	points.append(map(int,raw_input().split()))

count = 0

for i in xrange(len(points)-3):
	for j in xrange(i+1, len(points)-2):
		for k in xrange(j+1, len(points)-1):
			for l in xrange(k+1, len(points)):
				a = points[i]
				b = points[j]
				c = points[k]
				d = points[l]
				# print a, b, c, d
				tod = 0
				for perm in permutations((a, b, c, d)):
					ai, bj, ck, dl = list(perm)
					# print ai, bj, ck, dl
					# print "orientation a, b, c: ", orientation(ai, bj, ck), "orientation b,c,d:",orientation(bj,ck,dl)
					if orientation(ai, bj, ck) == 1 and orientation(bj,ck,dl) == -1 and isangle(ai,bj,ck) and isangle(bj,ck,dl):
						# print "Here"
						tod += 1
				count += tod/2

print count

# a=(2, 0)
# b=(2,2)
# c=(0,0)
# print orientation(a, b, c)

File: hackerrank/contests/week-of-code21/n-letter.cpp
#include <iostream>
#include <cmath>
#include <utility>

using namespace std;

#define pii pair<int, int>
#define PI 3.14159
#define mp make_pair

class Point{
public:
	int x, y;
	Point(){}
	Point(int a, int b){
		x = a;
		y = b;
	}
};

int orientation(Point a, Point b, Point c){
	float sigma = (float)(b.y - a.y) / (float)(b.x - a.x);
	float tou = (float)(c.y - b.y) / (float)(c.x - b.x);
	if (tou > sigma) return 1; // counter clock wise
	if (tou < sigma) return -1; // clock wise
	return 0; // linear
}

bool isangle(Point a, Point b, Point c){
	// returns whether angle is less than 90 degrees
	float Angle = acos(((b.x-a.x)*(c.x-b.x)+(b.y-a.y)*(c.y-b.y))/
			sqrt((b.x-a.x)*(b.y-a.y))*sqrt((c.x-b.x)*(c.y-b.y))); // this line is incorrect
	// here I've to divide by sqrt of squares
	Angle *= 180;
	cout << "Angle is " << Angle << endl;
	Angle /= PI;
	if (Angle > 90) return false;
	return true;
}

int main(void){
	Point a, b, c, d;
	a = Point(1, 1);
	b = Point(1, 2);
	c = Point(2,1);
	d = Point(2,2);
	cout << orientation(a, b, c) << " " << isangle(a, b, c) << endl;
	cout << orientation(b, c, d) << " " << isangle(b, c, d) << endl;
}